scp/scpInterpretationRequestInitiationAgent.cpp

```diff
@@ -0,0 +1,136 @@
+
+#include "scpKeynodes.hpp"
+
+#include <chrono>
+
+namespace scp
+{
+ScResult SCPInterpretationRequestInitiationAgent::DoProgram(ScElementaryEvent const & event, ScAction & action)
+{
+  auto const & startTime = std::chrono::high_resolution_clock::now();
+  ScAddr const & actionExpectedExecutionTimeLink = action.GetExpectedExecutionTimeInMilliseconds();
+  sc_uint32 actionExpectedExecutionTime = 0;
+  if (m_context.IsElement(actionExpectedExecutionTimeLink))
+    m_context.GetLinkContent(actionExpectedExecutionTimeLink, actionExpectedExecutionTime);
+
+  ScAddr const & agentProgram = GetScAgentProgram();
+  if (!m_context.IsElement(agentProgram))
+    action.FinishUnsuccessfully();
+
+  ScAddr const & scpParams = m_context.CreateNode(ScType::NodeConst);
+  ScAddr const & firstArgumentArc = m_context.CreateEdge(ScType::EdgeAccessConstPosPerm, scpParams, agentProgram);
+  m_context.CreateEdge(ScType::EdgeAccessConstPosPerm, Keynodes::rrel_1, firstArgumentArc);
+
+  ScAddr const & secondArgumentArc =
+      m_context.CreateEdge(ScType::EdgeAccessConstPosPerm, scpParams, event.GetConnector());
+  m_context.CreateEdge(ScType::EdgeAccessConstPosPerm, Keynodes::rrel_2, secondArgumentArc);
+
+  ScAction scpAction = m_context.GenerateAction(Keynodes::action_scp_interpretation_request);
+  scpAction.SetArguments(agentProgram, scpParams);
+
+  ScAddr const & authorArc = m_context.CreateEdge(ScType::EdgeDCommonConst, scpAction, Keynodes::abstract_scp_machine);
+  m_context.CreateEdge(ScType::EdgeAccessConstPosPerm, Keynodes::nrel_authors, authorArc);
+
+  if (actionExpectedExecutionTime == 0)
+    scpAction.InitiateAndWait();
+  else
+    scpAction.InitiateAndWait(
+        actionExpectedExecutionTime
+        - std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::high_resolution_clock::now() - startTime)
+              .count());
+  if (scpAction.IsFinishedSuccessfully())
+  {
+    ScAddr const & result = scpAction.GetResult();
+    if (!m_context.IsElement(result))
+    {
+      return action.FinishUnsuccessfully();
+    }
+    auto const & waiter =
+        m_context.GenerateConditionWaiter<ScEventAfterGenerateIncomingArc<ScType::EdgeAccessConstPosPerm>>(
+            result,
+            [](ScEventAfterGenerateIncomingArc<ScType::EdgeAccessConstPosPerm> const & subscribedEvent)
+            {
+              return subscribedEvent.GetArcSourceElement() == Keynodes::action_finished;
+            });
+    if (actionExpectedExecutionTime == 0)
+      waiter->Wait();
+    else
+      waiter->Wait(
+          actionExpectedExecutionTime
+          - std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::high_resolution_clock::now() - startTime)
+                .count());
+  }
+
+  return action.FinishSuccessfully();
+}
+
+ScAddr SCPInterpretationRequestInitiationAgent::GetScAgentProgram() const
+{
+  if (!m_context.IsElement(m_agentImplementationAddr))
+  {
+    SC_AGENT_LOG_ERROR("Agent implementation node is not valid");
```

Comment:
```suggestion
    SC_AGENT_LOG_ERROR("Agent implementation node is not valid.");
```

Reply:
I'll leave to the issue, I don't want to fix it now because I don't think that this '.' is crucial